src/rust/src/lib.rs

```diff
@@ -457,48 +439,47 @@ use rust_bridge::CxxFeeConfiguration;
 use rust_bridge::CxxLedgerEntryRentChange;
 use rust_bridge::CxxLedgerInfo;
 use rust_bridge::CxxRentFeeConfiguration;
+
 use rust_bridge::CxxTransactionResources;
 use rust_bridge::CxxWriteFeeConfiguration;
 use rust_bridge::FeePair;
 use rust_bridge::InvokeHostFunctionOutput;
 use rust_bridge::RustBuf;
-use rust_bridge::VersionNumPair;
-use rust_bridge::VersionStringPair;
-use rust_bridge::XDRHashesPair;
+use rust_bridge::SorobanVersionInfo;
 
 mod log;
 
 use crate::log::init_logging;
 
-// We have at least one, but possibly two, copies of soroban compiled
-// in to stellar-core. If we have two, ledgers that are exactly one
-// protocol _before_ the current (max-supported) protocol will run on
-// the `prev` copy of soroban. All others will run on the `curr` copy.
-// See `invoke_host_function` below.
+// We have multiple copies of soroban linked into stellar-core here. This is
+// accomplished using an adaptor module -- contract.rs -- mounted multiple times
+// into the same outer crate, inside different modules soroban_p21, soroban_p22,
+// etc. each with its own local binding for the external crate soroban_env_host.
+// The contract.rs module imports soroban_env_host from `super` which means each
+// instance of it sees a different soroban. This is a bit of a hack and only
+// works when the soroban versions all have a compatible _enough_ interface to
+// all be called from "the same" contract.rs.
 
 #[path = "."]
-mod soroban_curr {
-    pub(crate) use soroban_env_host_curr as soroban_env_host;
+mod soroban_p22 {
+    pub(crate) use soroban_env_host_p22 as soroban_env_host;
 
     pub(crate) mod contract;
 }
 
-#[cfg(feature = "soroban-env-host-prev")]
 #[path = "."]
-mod soroban_prev {
-    pub(crate) use soroban_env_host_prev as soroban_env_host;
+mod soroban_p21 {
+    pub(crate) use soroban_env_host_p21 as soroban_env_host;
 
     pub(crate) mod contract;
 }
 
-#[cfg(feature = "soroban-env-host-prev")]
-pub fn compiled_with_soroban_prev() -> bool {
-    true
-}
+// We alias the latest soroban as soroban_curr to help reduce churn in code
+// that's just always supposed to use the latest.
+use soroban_p22 as soroban_curr;
 
-#[cfg(not(feature = "soroban-env-host-prev"))]
 pub fn compiled_with_soroban_prev() -> bool {
```

Comment:
I think it should be safe to just remove this from here and Upgrades, is there any particular reason to keep it?

Reply:
Agreed. I think this check was probably overkill. We could write a check that takes two protocol numbers and confirms that they are being handled by two different sorobans I guess? I don't feel strongly.